src/lender/tranche/v3/tranche.sol

```diff
@@ -0,0 +1,169 @@
+    uint supplyFulfillment;
+    // tokenPrice after end of epoch
+    uint tokenPrice; 
+    mapping (address => uint) supplyCurrencyAmount;
+    mapping (address => uint) redeemTokenAmount;
+}
+
+ERC20Like public currency;
+ERC20Like public token;
+TickerLike public ticker;
+address self;
+
+constructor(address currency_, address token_, address ticker_) public {
+    wards[msg.sender] = 1;
+    token = ERC20Like(token_);
+    currency = ERC20Like(currency_);
+    ticker = TickerLike(ticker_);
+    self = address(this);
+}
+
+function supplyCurrencyAmount(uint epochID, address addr) public view returns (uint) {
+    Epoch storage epoch = epochs[epochID];
+    return epoch.supplyCurrencyAmount[addr];
+}
+
+function redeemTokenAmount(uint epochID, address addr) public view returns (uint) {
+    Epoch storage epoch = epochs[epochID];
+    return epoch.redeemTokenAmount[addr];
+}
+
+function balance() external view returns (uint) {
+    return currency.balanceOf(self);
+}
+
+function tokenSupply() external view returns (uint) {
+    return token.totalSupply();
+}
+
+function depend(bytes32 contractName, address addr) public auth {
+    if (contractName == "token") { token = ERC20Like(addr); }
+    if (contractName == "currency") { currency = ERC20Like(addr); }
+    if (contractName == "ticker") { ticker = TickerLike(addr); }
+    else revert();
+}
+
+// supplyOrder function can be used to place or revoke an supply 
+function supplyOrder(uint epochID, uint supplyAmount) public {
+    require((epochID >= ticker.currentEpoch()), "epoch-already-over");
+    uint currentSupplyAmount = epochs[epochID].supplyCurrencyAmount[msg.sender];
+    epochs[epochID].supplyCurrencyAmount[msg.sender] = supplyAmount;
+    epochs[epochID].totalSupply = safeAdd(safeSub(epochs[epochID].totalSupply, currentSupplyAmount), supplyAmount);
+    if (supplyAmount > currentSupplyAmount) {
+        uint delta = safeSub(supplyAmount, currentSupplyAmount);
+        require(currency.transferFrom(msg.sender, self, delta), "currency-transfer-failed");
+        return;
+    } 
+    uint delta = safeSub(currentSupplyAmount, supplyAmount);
+    if (delta > 0) {
+       require(currency.transferFrom(self, msg.sender, delta), "currency-transfer-failed");
+    }
+}
+
+// redeemOrder function can be used to place or revoke a redeem
+function redeemOrder(uint epochID, uint redeemAmount) public {
+    require((epochID >= ticker.currentEpoch()), "epoch-already-over");
+    
+    uint currentRedeemAmount = epochs[epochID].redeemTokenAmount[msg.sender];
+    epochs[epochID].redeemTokenAmount[msg.sender] = redeemAmount;
+    epochs[epochID].totalRedeem = safeAdd(safeSub(epochs[epochID].totalRedeem, currentRedeemAmount), redeemAmount);
+    
+     if (redeemAmount > currentRedeemAmount) {
+        uint delta = safeSub(redeemAmount, currentRedeemAmount);
+        require(token.transferFrom(msg.sender, self, delta), "token-transfer-failed");
+        token.burn(self, redeemAmount);
```

Comment:
We should burn and mint only in the epoch execute because it impacts the tokenPrice.

Assume Alice and Bob both invest 50 DAI for 1.00DAI token Price. Pool would be 100 DAI for 100 Token.

If Alice would now make an redeemOrder we would burn the 50 Token and the token price would be 100/50 = 2.0.

Alice would receive 100 DAI for her 50 token.